refactor: Replace boost::bind with std::bind

ken2812221 opened the pull request:
Replace boost::bind with std::bind

- In `src/rpc/server.cpp`, replace `std::transform` with simple loop.
- In `src/validation.cpp`, store the `boost::signals2::connection` object and use it to disconnect.
- In `src/validationinterface.cpp`, use 2 map to store the `boost::signals2::scoped_connection` object.

fanquake commented:
Concept ACK

After this `boost::bind` will still be used by server, scheduler and validation (and associated tests). @ken2812221 Is there a particular reason you haven't removed that usage?

practicalswift commented:
Concept ACK, but please remove `using namespace std::placeholders;`. Reference `_N` using the fully specified `std::placeholders::_N`.

See developer notes:

> - Don't import anything into the global namespace (`using namespace ...`). Use fully specified types such as `std::string`.

ken2812221 commented:
>Is there a particular reason you haven't removed that usage?

Those instances cannot be replaced simply `boost` -> `std`. I'll try to get rid of those.

>please remove `using namespace std::placeholders;`

Will do

MarcoFalke commented:
I believe it is acceptable to be using `using namespace std::placeholders` here, since the alternative would be extremely verbose.

practicalswift commented:
@MarcoFalke I'm not sure I agree. The verbosity could help remind us that we should use lambdas instead of ugly `std::bind` (modulo the few places where that is not technically possible in C++11), no? :-)

ken2812221 commented:
@practicalswift Since @MarcoFalke think that `using namespace std::placeholders;` is acceptable,  I will wait for more comments from other people and decide what to do.

ken2812221 commented:
@practicalswift Good point! I'll try to use lambda to replace bind

Empact commented on src/validationinterface.cpp:
```diff
@@ -73,27 +82,35 @@ CMainSignals& GetMainSignals()
+    connections.Broadcast = g_signals.m_internals->Broadcast.connect(std::bind(&CValidationInterface::ResendWalletTransactions, pwalletIn, _1, _2));
+    connections.BlockChecked = g_signals.m_internals->BlockChecked.connect(std::bind(&CValidationInterface::BlockChecked, pwalletIn, _1, _2));
+    connections.NewPoWValidBlock = g_signals.m_internals->NewPoWValidBlock.connect(std::bind(&CValidationInterface::NewPoWValidBlock, pwalletIn, _1, _2));
+    g_signals.m_wallet_connections.emplace(pwalletIn, connections);
```
Can `std::move` `connections`

Empact commented on src/validationinterface.cpp:
```diff
@@ -73,27 +82,35 @@ CMainSignals& GetMainSignals()
+    connections.ChainStateFlushed.disconnect();
+    connections.Broadcast.disconnect();
+    connections.BlockChecked.disconnect();
+    connections.NewPoWValidBlock.disconnect();
```
nit: could extract these as a method on `ValidationInterfaceConnections` where it would be more clear whether the method was comprehensive in its disconnecting.

Empact commented on src/validationinterface.cpp:
```diff
@@ -73,27 +82,35 @@ CMainSignals& GetMainSignals()
+    connections.ChainStateFlushed = g_signals.m_internals->ChainStateFlushed.connect(std::bind(&CValidationInterface::ChainStateFlushed, pwalletIn, _1));
+    connections.Broadcast = g_signals.m_internals->Broadcast.connect(std::bind(&CValidationInterface::ResendWalletTransactions, pwalletIn, _1, _2));
+    connections.BlockChecked = g_signals.m_internals->BlockChecked.connect(std::bind(&CValidationInterface::BlockChecked, pwalletIn, _1, _2));
+    connections.NewPoWValidBlock = g_signals.m_internals->NewPoWValidBlock.connect(std::bind(&CValidationInterface::NewPoWValidBlock, pwalletIn, _1, _2));
```
nit: could extract as a method on `ValidationInterfaceConnections`

promag commented:
Concept ACK.

MarcoFalke commented on src/validationinterface.h:
```diff
@@ -6,10 +6,13 @@
 #define BITCOIN_VALIDATIONINTERFACE_H
 
+#include <boost/signals2/signal.hpp>
+
```
I believe this will slow down the build significantly. So tend to NACK here.

ken2812221 replied:
Fixed

MarcoFalke commented:
Could you sum up the differences between `std::bind` and lambda in the first comment for the convenience of the reviewers? It might not be obvious to everyone why lambdas should be preferred over `std::bind` everywhere.

practicalswift commented:
@MarcoFalke Stephan T. Lavavej (standard library maintainer Visual C++) sums up the `std::bind` issues/gotchas here: https://www.youtube.com/watch?v=zt7ThwVfap0&t=1721

His recommendations:

> * Avoid using `bind()`
> * Use lambdas, especially generic lambdas

theuni commented:
Please break this up into logical commits for review.
boost::bind -> std::bind makes sense to me. Concept ACK there for sure.

std::bind -> lambda seems more like a style/preference change. I prefer lambdas too, but I don't think that justifies the churn here. Maybe when lambdas are more robust and provide a more clear win with c++14/c++17?

fanquake commented:
Travis linter is failing with:
```
Running script for: fcba334f0d21161354f168a081d95a63b82fa4c5
ALL_FILES=$(git ls-files -- '*.cpp' '*.h')
for i in 1 2 3 4 5; do
    sed -i "s/ _${i}/ std::placeholders::_${i}/g" ${ALL_FILES}
done
fatal: ambiguous argument '5': unknown revision or path not in the working tree.
```

ken2812221 commented:
Update: In this PR, I do boost::bind to std::bind only because there are some disputes about lambda.

practicalswift commented:
@ken2812221 Makes sense! Getting rid of `boost/bind.hpp` is a very welcome first step :-)

Empact commented:
Concept ACK, the callers should include `<functional>`

ryanofsky commented on src/validation.cpp:
```diff
@@ -2368,14 +2368,15 @@ class ConnectTrace {
 
     ~ConnectTrace() {
-        pool.NotifyEntryRemoved.disconnect(boost::bind(&ConnectTrace::NotifyEntryRemoved, this, _1, _2));
+        m_connNotifyEntryRemoved.disconnect();
```
Could drop this disconnect line (and the whole destructor) by replacing ` boost::signals2::connection` with `boost::signals2::scoped_connection` above.

ryanofsky commented on src/validationinterface.cpp:
```diff
@@ -37,6 +37,8 @@ struct MainSignalsInstance {
 
 static CMainSignals g_signals;
+static std::unordered_map<CTxMemPool*, boost::signals2::connection> g_connNotifyEntryRemoved;
+static std::unordered_map<CValidationInterface*, std::array<boost::signals2::connection, 9>> g_connMainSignals;
```
It seems like these two maps would most naturally be `MainSignalsInstance` members. I don't think it is good to arbitrary divide the validationinterface state up between `g_signals->m_internals` and these new global variables (unless there's a specific reason for doing this that I'm not seeing).

ken2812221 replied:
I was trying to do this however MainSignalsInstance is not created when `RegisterWithMempoolSignals()` called. So I only move `g_connMainSignals` into a member of `MainSignalsInstance`. Keeping `g_connNotifyEntryRemoved` as a static global variable.

ryanofsky commented on src/validationinterface.cpp:
```diff
@@ -72,27 +78,25 @@ CMainSignals& GetMainSignals()
-    g_signals.m_internals->NewPoWValidBlock.disconnect(boost::bind(&CValidationInterface::NewPoWValidBlock, pwalletIn, _1, _2));
+    const auto iter = g_connMainSignals.find(pwalletIn);
+    if (iter == g_connMainSignals.end()) return;
+    for (const boost::signals2::connection& conn : iter->second) {
```
I think if you used `scoped_connection` instead of `connection` types you could get rid of this for loop.

Also, since this for loop is the only thing that appears to iterate over `std::array<9>` objects, it might be a good idea to replace `std::array` with a plain struct if you do eliminate this loop. (I noticed you used a `ValidationInterfaceConnections` struct in a previous version of this PR).

ken2812221 replied:
That idea looks good to me. I'll take it.

ryanofsky reviewed (approved):
utACK d4a4f5729f0faa3bc51c6f162a10812c9703d5a5. I also think it'd be good to switch to lambdas in the future.

sipa commented:
Can this be implemented as a scripted diff? It looks relatively straightforward.

ken2812221 committed:
refactor: Use boost::scoped_connection in signal/slot, also prefer range-based loop instead of std::transform

ken2812221 committed:
scripted-diff: Replace boost::bind with std::bind

-BEGIN VERIFY SCRIPT-
for j in $(seq 1 5)
do
    sed -i "s/ _${j}/ std::placeholders::_${j}/g" $(git grep --name-only " _${j}" -- '*.cpp' '*.h')
done
sed -i "s/boost::bind/std::bind/g" $(git grep --name-only boost::bind -- '*.cpp' '*.h')
sed -i "s/boost::ref/std::ref/g" $(git grep --name-only boost::ref -- '*.cpp' '*.h')
sed -i '/boost\/bind/d' $(git grep --name-only boost/bind)
-END VERIFY SCRIPT-

ken2812221 commented:
@sipa Done. Apparently I can't use i as the script's variable.

ryanofsky commented on src/validationinterface.cpp:
```diff
@@ -32,11 +44,13 @@ struct MainSignalsInstance {
 };
 
 static CMainSignals g_signals;
+static std::unordered_map<CTxMemPool*, boost::signals2::scoped_connection> g_connNotifyEntryRemoved;
```
Following up on https://github.com/bitcoin/bitcoin/pull/13743#discussion_r224988537, I think it would be good to add a comment saying that this has to a separate global instead of a member of MainSignalsInstance, because RegisterWithMempoolSignals is currently called before RegisterBackgroundSignalScheduler, so MainSignalsInstance hasn't been created yet.

Init code is more messy than it needs to be and I think saying what current dependencies are will make it easier to clean up later.

ryanofsky reviewed (approved):
utACK 52532ee43720520696d452424daad213b50aef1f. Changes since last review are adding scripted diff, removing one global, and switching to scoped signal connections.

> Apparently I can't use `${i}` as the script's variable.

It looks like this happens because [commit-script-check.sh](https://github.com/bitcoin/bitcoin/blob/master/test/lint/commit-script-check.sh) uses `$i` and calls `eval "$SCRIPT"` without a subshell. It would be nice to fix this to prevent confusion and ensure the diff checking is correct.

ryanofsky reviewed (approved):
utACK cb53b825c26af6e628ba88d72b2000e75bedbbc6. Only change is last review is new comment.

sipa commented:
utACK cb53b825c26af6e628ba88d72b2000e75bedbbc6

ryanofsky commented:
Can this be merged, or does it need more review? I see review from @theuni was requested in github, but I don't know if it's necessary given the other reviews since this change is pretty straightforward.

fanquake commented:
utACK cb53b82

practicalswift commented:
utACK cb53b825c26af6e628ba88d72b2000e75bedbbc6
